File: src/ast/tree/statement/native.ts
import type { Pass } from "../../../pass/pass";
import type { Block } from "../../node/block";
import { ExecutionResult, Statement } from "../../node/statement";

export class Native extends Statement {
    constructor(
        public func: (native: Native) => void,
        parent: Block | null
    ) {
        super(parent);
    }

    public execute(): ExecutionResult {
        this.func(this);
        return ExecutionResult.ContinueEvaluation;
    }

    public visit(pass: Pass): void {
        //pass.visitNative(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/statement/Native.kt
package me.honkling.miniscript.parser.ast.statement

import me.honkling.miniscript.MiniScript
import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.function.Function
import me.honkling.miniscript.pass.Pass

fun nativeBlock(miniScript: MiniScript, function: Function, runnable: Native.() -> Unit): Block {
    val block = Block(miniScript, mutableListOf(), function)
    val statement = Native(runnable, block)
    block.statements += statement
    return block
}

class Native(val block: Native.() -> Unit, parent: Block) : Statement(parent) {
    override fun execute(): ExecutionResult {
        block(this)
        return ExecutionResult.ContinueExecution
    }

    override fun accept(pass: Pass) {
        pass.visitNative(this)
    }
}